Decouple Place and Expression in ExprUseVisitor Delegate

null-sleep opened the pull request:
Modifies the `Delegate` interface to take a `place: Place` and an `expr_id: HirID`, where `expr_id` is pointed to in diagnostics.

Issue: https://github.com/rust-lang/project-rfc-2229/issues/20

Example:

```
let [a, ...]      =     arr;
^^Expr1^^ = ^Expr2^
```

Here `arr` is moved because of the binding created in Expr1. However, in diagnostics we point to Expr1 with the message `arr` was moved, which can be confusing. With this change, the diagnostics will point to Expr2 for patterns.

We have tried to previously achieve something similar before https://github.com/rust-lang/rust/pull/75933/files#diff-1f18144176a6e44f905262991bed987098348cdbd688ac76f8c915ec91b7fd58R316-R319

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sexxi-goose/rust/27)
<!-- Reviewable:end -->

roxelo commented on src/test/ui/borrowck/borrowck-closures-slice-patterns.stderr:
```diff
@@ -75,7 +75,7 @@ LL | fn arr_box_by_move(x: Box<[String; 3]>) {
    |             -- value moved into closure here
 LL |         let [y, z @ ..] = *x;
-   |                            - variable moved due to use in closure
+   |                           -- variable moved due to use in closure
```
Is this the only test case impacted? From my understanding of the changes, it seems like at least new test cases should be introduced to confirm the behavior is now different (if none of the other one are currently impacted).

arora-aman replied:
this change goes in before the new behavior

arora-aman replied:
Also the new behavior doesn't affect things yet

null-sleep committed:
Provide diagnostic suggestion in ExprUseVisitor Delegate

The [Delegate
trait](https://github.com/rust-lang/rust/blob/981346fc07dd5ef414c5b1b21999f7604cece006/compiler/rustc_typeck/src/expr_use_visitor.rs#L28-L38)
currently use `PlaceWithHirId` which is composed of Hir `Place` and the
corresponding expression id.

Even though this is an accurate way of expressing how a Place is used,
it can cause confusion during diagnostics.

Eg:

```
let arr : [String; 5];

let [a, ...]     =   arr;
 ^^^ E1 ^^^      =  ^^E2^^
 ```

 Here `arr` is moved because of the binding created E1. However, when we
 point to E1 in diagnostics with the message `arr` was moved, it can be
 confusing.  Rather we would like to report E2 to the user.

 Closes: https://github.com/rust-lang/project-rfc-2229/issues/20

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -278,12 +278,13 @@ struct InferBorrowKind<'a, 'tcx> {
     fn adjust_upvar_borrow_kind_for_consume(
         &mut self,
-        place_with_id: &PlaceWithHirId<'tcx>,
+        place: &PlaceWithHirId<'tcx>,
```
```suggestion
        place_with_id: &PlaceWithHirId<'tcx>,
```

null-sleep replied:
👍 Missed a few, fixing

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -343,10 +337,17 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
 
     /// Indicates that `place_with_id` is being directly mutated (e.g., assigned
-    /// to). If the place is based on a by-ref upvar, this implies that
+    ///  to). If the place is based on a by-ref upvar, this implies that
```
extra space

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -500,29 +504,38 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
-        self.adjust_upvar_borrow_kind_for_consume(place_with_id, mode);
+    fn consume(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
`place_with_id`

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -500,29 +504,38 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
-        debug!("borrow(place_with_id={:?}, bk={:?})", place_with_id, bk);
+    fn borrow(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
`place_with_id`

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
+    // looks confusing. Instead we prefer to report the discriminant (rhs)
+    fn consume(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
`place_with_id`

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
+    // looks confusing. Instead we prefer to report the discriminant (rhs)
+    fn borrow(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
`place_with_id`

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -116,11 +140,11 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
 
-    fn delegate_consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>) {
-        debug!("delegate_consume(place_with_id={:?})", place_with_id);
+    fn delegate_consume(&mut self, place: &PlaceWithHirId<'tcx>, diag_expr_id: hir::HirId) {
```
`place_with_id`

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -590,11 +625,10 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
-        place_with_id.place.ty(),
-        mc.tcx().hir().span(place_with_id.hir_id),
-    ) {
+    if !mc.type_is_copy_modulo_regions(place_with_id.place.ty(), mc.tcx().hir().span(diag_expr_id))
```
Honestly not sure if we should be using `diag_expr_id` or `place_with_id.hir_id`

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -476,13 +500,17 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
             adjustment::AutoBorrow::Ref(_, m) => {
-                self.delegate.borrow(base_place, ty::BorrowKind::from_mutbl(m.into()));
+                self.delegate.borrow(
+                    &base_place,
```
```suggestion
                    base_place,
```

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -476,13 +500,17 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                 debug!("walk_autoref: expr.hir_id={} base_place={:?}", expr.hir_id, base_place);
 
-                self.delegate.borrow(base_place, ty::BorrowKind::from_mutbl(m));
+                self.delegate.borrow(&base_place, base_place.hir_id, ty::BorrowKind::from_mutbl(m));
```
```suggestion
                self.delegate.borrow(base_place, base_place.hir_id, ty::BorrowKind::from_mutbl(m));
```

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -525,19 +553,22 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                     let def = Res::Local(canonical_id);
                     if let Ok(ref binding_place) = mc.cat_res(pat.hir_id, pat.span, pat_ty, def) {
-                        delegate.mutate(binding_place);
+                        delegate.mutate(&binding_place, binding_place.hir_id);
```
[](url)
```suggestion
                        delegate.mutate(binding_place, binding_place.hir_id);
```

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -525,19 +553,22 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                         ty::BindByReference(m) => {
                             let bk = ty::BorrowKind::from_mutbl(m);
-                            delegate.borrow(place, bk);
+                            delegate.borrow(&place, discr_place.hir_id, bk);
```
```suggestion
                            delegate.borrow(place, discr_place.hir_id, bk);
```

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -525,19 +553,22 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
+                            let mode = copy_or_move(mc, &place);
                             debug!("walk_pat binding consuming pat");
-                            delegate.consume(place, mode);
+                            delegate.consume(&place, discr_place.hir_id, mode);
```
```suggestion
                            delegate.consume(place, discr_place.hir_id, mode);
```

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -116,11 +140,11 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
 
-        let mode = copy_or_move(&self.mc, place_with_id);
-        self.delegate.consume(place_with_id, mode);
+        let mode = copy_or_move(&self.mc, &place_with_id);
```
```suggestion
        let mode = copy_or_move(&self.mc, place_with_id);
```

arora-aman reviewed (changes requested):
No need to reborrow

null-sleep committed:
Address Pr comments regarding docstrings
